teps/0150-displayname-in-matrix.md

```diff
@@ -0,0 +1,304 @@
+    params:
+      - name: platform
+        value: $(params.platforms[*])
+      - name: browser
+        value: $(params.browsers[*])
+  taskRef:
+    name: run-tests
+```
+
+#### `matrix` with `include`
+
+```yaml
+- name: platforms-and-browsers
+  matrix:
+    displayName: "Platforms and Browsers: $(params.platform) and $(params.browser)"
+    include:
+      - params:
+          - name: platform
+            value: "linux"
+          - name: browser
+            value: "chrome"
+      - params:
+          - name: platform
+            value: "mac"
+          - name: browser
+            value: "safari"
+  taskRef:
+    name: run-tests
+```
+
+#### `matrix` with `params` which depends on the `taskResult`
+
+```yaml
+- name: run-tests
+  matrix:
+    displayName: "Run Tests - $(params.VARIANT)"
+    params:
+      - name: VARIANT
+        value: $(tasks.calculate-vars.results.variations)
+  taskRef:
+    name: run-tests-task
+```
+
+#### constant `displayName`
+
+```yaml
+- name: generate-report
+  matrix:
+    displayName: "Generate Report"
+    params:
+      - name: type
+        value: $(tasks.discover-report-types.results.types)
+  taskRef:
+    name: generate-report-task
+```
+
+The resolved `matrix.displayName` will be available as part of `pipelineRun.status.childReferences` with each `taskRun`.
+This allows Tekton Dashboard to consume `matrix.displayName` whenever available. The `displayName` in
+`pipelineRun.status.childReferences` for a `taskRun` if it is a part of `matrix`.
+
+The `pipeline` with the second example of specifying `displayName` with `include` results in the following
+`pipelineRun.status.childReferences`:
+
+```json
+[
+  {
+    "apiVersion": "tekton.dev/v1",
+    "kind": "TaskRun",
+    "name": "matrixed-pr-vcx79-task-0",
+    "pipelineTaskName": "task-0"
+  },
+  {
+    "apiVersion": "tekton.dev/v1",
+    "displayName": "Platforms and Browsers: linux and chrome",
+    "kind": "TaskRun",
+    "name": "matrixed-pr-vcx79-platforms-and-browsers-0",
+    "pipelineTaskName": "platforms-and-browsers"
+  },
+  {
+    "apiVersion": "tekton.dev/v1",
+    "displayName": "Platforms and Browsers: mac and safari",
+    "kind": "TaskRun",
+    "name": "matrixed-pr-vcx79-platforms-and-browsers-1",
+    "pipelineTaskName": "platforms-and-browsers"
+  }
+]
+```
+
+`matrix.displayName` is not enabled by default by can be enabled with a
+[feature flag](https://github.com/tektoncd/pipeline/blob/main/config/config-feature-flags.yaml) `matrix-extensions` and
+setting it to `displayName`.
+
+### `tasks[].displayName` and `tasks[].matrix.displayName`
+
+The `spec.tasks[].displayName` was introduced in [TEP-0047](0047-pipeline-task-display-name.md) which originally allowed
+any human-readable plain text and later [enhanced](https://github.com/tektoncd/pipeline/pull/7273) to support using
+`params` to allow runtime information in the `displayName`. The `spec.tasks[].displayName` and `spec.tasks[].matrix.displayName`
```

Comment:
Supporting `displayName` is 2 places seems a bit odd but referencing matrix params in `spec.tasks[].displayName` also doesn't seem right, so I'm interested in feedback from others on this. Perhaps having 2 `displayName` fields is the best compromise.

Reply:
If we only have the single `spec.tasks[].displayName` I think we'd be best to introduce a new variable substitution specific to the matrix params. Something like 
```
displayName: "Platforms and Browsers: $(matrix.params.platform) and $(matrix.params.browser)"
```

Reply:
Since both `spec.tasks[].displayName` and `spec.tasks[].matrix.displayName` would be defined at pipeline level, I tend to agree with @AlanGreene that having two places feels redundant. 
I like @AverageMarcus proposal of having matrix specific variables to display in `spec.tasks[].displayName`, that way we would avoid the extra API field.

Reply:
thank you @AlanGreene, @AverageMarcus, and @afrittoli for the input 🙏 

I think its a question of introducing a new way of referencing `params` i.e. `matrix.params.<param>` v/s introducing a new field. I am open to reusing the existing field `spec.tasks[].displayName` instead of introducing a new field. But wondering is it possible? 

The purpose of introducing a new field is a new functionality - `matrix.displayName` will be part of the `childReferences` such that the clients/CLI/dashboard can rely on `displayName` instead of `pipelineTaskName` from the `childReferences` to get a unique/parameterized name.

To reuse the existing field, `tasks[].displayName` using `matrix` specific substitution, are we aligned on the rest of the proposal to make it part of the `childReferences`?

So far, `tasks[].displayName` is capable of resolving pipeline params:

```
kind: Pipeline
spec:
  params:
    - name: git-repo
  tasks:
    - name: scan
      displayName: "Code Scan for $(params.git-repo)"
      taskRef:
        name: sonar-scan
```

We **do not** have a mechanism in place to resolve and replace `pipelineTask` params:

```
kind: Pipeline
spec:
  tasks:
    - name: scan
      params:
        - name: git-repo
           value: "my-git-repo"
      displayName: "Code Scan for $(params.git-repo)"
      taskRef:
        name: sonar-scan
```

We might be able to change this but just disclosing it here for the awareness.

Reply:
> are we aligned on the rest of the proposal to make it part of the childReferences?

Yes I think this makes sense. Ideally the displayName would always be available to clients in childReferences, regardless of how/where it's set, what substitutions it uses, etc.

As for the issue with pipelineTask params, perhaps that could be addressed in a separate proposal if there's a need. If I understand correctly it's an existing limitation and not affected by the changes proposed here.

Reply:
Thank you @AlanGreene 👍 

In case of always including `displayName` in the `childReferences`, it is doable to fully resolve `displayName` with `pipelineTask`-specific params in addition to the task results. A task only makes into the `childReferences` once it is done and all the params/results are resolved and available for the further resolution.

How does the dashboard consume `displayName` today? Can we switch to using it through `childReferences` such that it has the ultimate resolved text?

Reply:
> Can we switch to using it through childReferences such that it has the ultimate resolved text?

Yes that's the plan